kyo-data/shared/src/test/scala/kyo/RecordTest.scala

```diff
@@ -0,0 +1,470 @@
+            assert(record.`user-name` == "Alice")
+            assert(record._internal == 42)
+        }
+
+        "null values" in {
+            val record = "name" ~ null
+            assert(record.name == null)
+        }
+
+        "empty case class" in {
+            case class Empty()
+            val record = Record.fromProduct(Empty())
+            assert(record.size == 0)
+        }
+
+        "case class with 22+ fields" in {
+            case class Large(
+                f1: Int,
+                f2: Int,
+                f3: Int,
+                f4: Int,
+                f5: Int,
+                f6: Int,
+                f7: Int,
+                f8: Int,
+                f9: Int,
+                f10: Int,
+                f11: Int,
+                f12: Int,
+                f13: Int,
+                f14: Int,
+                f15: Int,
+                f16: Int,
+                f17: Int,
+                f18: Int,
+                f19: Int,
+                f20: Int,
+                f21: Int,
+                f22: Int,
+                f23: Int
+            )
+            val large  = Large(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23)
+            val record = Record.fromProduct(large)
+            assert(record.size == 23)
+            assert(record.f1 == 1)
+            assert(record.f23 == 23)
+        }
+    }
+
+    "modification" - {
+        "adding new fields" in {
+            val record1 = "name" ~ "Charlie"
+            val record2 = record1 & "age" ~ 25
+            assert(record2.name == "Charlie")
+            assert(record2.age == 25)
+        }
+
+        "updating existing fields" in {
+            val record1 = "name" ~ "Charlie" & "age" ~ 25
+            val record2 = record1 & "age" ~ 26
+            assert(record2.name == "Charlie")
+            assert(record2.age == 26)
+        }
+    }
+
+    "combining records" - {
+        "merge two records" in {
+            val record1 = "name" ~ "Eve" & "age" ~ 30
+            val record2 = "city" ~ "Paris" & "country" ~ "France"
+            val merged  = record1 & record2
+
+            assert(merged.name == "Eve")
+            assert(merged.age == 30)
+            assert(merged.city == "Paris")
+            assert(merged.country == "France")
+        }
+
+        "merge with overlapping fields" in {
+            val record1 = "name" ~ "Eve" & "age" ~ 30
+            val record2 = "age" ~ 31 & "city" ~ "Paris"
+            val merged  = record1 & record2
+
+            assert(merged.name == "Eve")
+            assert(merged.age == 31)
+            assert(merged.city == "Paris")
+        }
+
+        "merge with type conflicts" in {
+            val record1 = "value" ~ "string"
+            val record2 = "value" ~ 42
+            val merged  = record1 & record2
+            assert((merged.value: Int) == 42)
+            assert((merged.value: String) == "string")
```

Comment:
This seems like a foot-gun. Are you sure you want to support having multiple types? What's the behavior if a type annotation is not provided explicitly? Does it really make sense to support multiple values of different types for a given field name?

Reply:
I actually didn't want to have this behavior. Ideally, `"field1" ~ String & "field2" ~ Int` should be inferred as `"field1" ~ (String | Int)` but I tried a few times to play with the variance and couldn't make it work. The alternative I found was allowing duplicate names and using tags to disambiguate. It's still safe to have fields with the same name and different types since all access is based on both components.

Reply:
About the behavior in case the compiler doesn’t infer a specific result type, the first matching field and its specific type is automatically selected. I agree it’s a rough edge. I wonder if it’d be possible to disallow name conflicts via macros.